frontend/src/container/ActivityFeed.jsx

```diff
@@ -0,0 +1,228 @@
+  sortActivityList
+} from '../util/activity.js'
+import {
+  FETCH_CONFIG,
+  PAGE
+} from '../util/helper.js'
+import {
+  getNotificationList,
+  getWorkspaceDetail
+} from '../action-creator.async.js'
+import {
+  setWorkspaceActivityList,
+  setWorkspaceActivityNextPage,
+  setBreadcrumbs,
+  setHeadTitle,
+  newFlashMessage,
+  setWorkspaceDetail
+} from '../action-creator.sync.js'
+
+import TabBar from '../component/TabBar/TabBar.jsx'
+import ContentWithPreviewActivity from '../component/Activity/ContentWithPreviewActivity.jsx'
+import ContentWithoutPreviewActivity from '../component/Activity/ContentWithoutPreviewActivity.jsx'
+import MemberActivity from '../component/Activity/MemberActivity.jsx'
+
+require('../css/ActivityFeed.styl')
+
+const ACTIVITY_COUNT_PER_PAGE = NUMBER_RESULTS_BY_PAGE
+const NOTIFICATION_COUNT_PER_REQUEST = ACTIVITY_COUNT_PER_PAGE
+const ENTITY_TYPE_COMPONENT_CONSTRUCTOR = new Map([
+  [TLM_ET.CONTENT, (activity) => {
+    return activity.newestMessage.fields.content.content_type === CONTENT_TYPE.FOLDER
+      ? <ContentWithoutPreviewActivity activity={activity} key={activity.id} />
+      : <ContentWithPreviewActivity activity={activity} key={activity.id} />
+  }],
+  [TLM_ET.SHAREDSPACE_MEMBER, (activity) => <MemberActivity activity={activity} key={activity.id} />],
+  [TLM_ET.SHAREDSPACE_SUBSCRIPTION, (activity) => <MemberActivity activity={activity} key={activity.id} />]
+])
+
+export class ActivityFeed extends React.Component {
+  constructor (props) {
+    super(props)
+    props.registerGlobalLiveMessageHandler(this.updateActivityList.bind(this))
+  }
+
+  componentDidMount () {
+    this.loadWorkspaceDetail()
+    this.loadActivities(ACTIVITY_COUNT_PER_PAGE, true)
+    this.setHeadTitle()
+    this.buildBreadcrumbs()
+  }
+
+  componentDidUpdate (prevProps) {
+    if (prevProps.workspaceId === this.props.workspaceId) return
+    this.loadWorkspaceDetail()
+    this.loadActivities(ACTIVITY_COUNT_PER_PAGE, true)
+    this.setHeadTitle()
+    this.buildBreadcrumbs()
+  }
+
+  async updateActivityList (data) {
+    const { props } = this
+    const updatedActivityList = await addMessageToActivityList(data, props.workspaceActivity.list, FETCH_CONFIG.apiUrl)
+    props.dispatch(setWorkspaceActivityList(updatedActivityList))
+  }
+
+  sortActivityList () {
+    const { props } = this
+    const updatedActivityList = sortActivityList(props.workspaceActivity.list)
+    props.dispatch(setWorkspaceActivityList(updatedActivityList))
+  }
+
+  async loadActivities (minActivityCount, resetList = false) {
+    const { props } = this
+    let activityList = resetList ? [] : props.workspaceActivity.list
+    let hasNextPage = resetList ? true : props.workspaceActivity.hasNextPage
+    let nextPageToken = resetList ? '' : props.workspaceActivity.nextPageToken
+    while (hasNextPage && activityList.length < minActivityCount) {
+      const messageListResponse = await props.dispatch(getNotificationList(
```

Comment:
If you `await` inside the `while`, it will force the loop to wait for the api to have responded until each iterations. It might create a long loading time by blocking parallelization 

I don't know if it is required

Reply:
It is needed as we want to get several pages of notifications and to get a page we need the previous one (for its `nextPageToken`). The while is needed because we don't know how many notifications pages we'll need to build the needed activities count.

Reply:
I see, it's fine then